sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/rel/BeamEnumerableConverter.java

```diff
@@ -83,12 +81,25 @@ public Result implement(EnumerableRelImplementor implementor, Prefer prefer) {
     final RelDataType rowType = getRowType();
     final PhysType physType =
         PhysTypeImpl.of(implementor.getTypeFactory(), rowType, prefer.preferArray());
-    final Expression options = implementor.stash(this.options, PipelineOptions.class);
     final Expression node = implementor.stash((BeamRelNode) getInput(), BeamRelNode.class);
-    list.add(Expressions.call(BeamEnumerableConverter.class, "toEnumerable", options, node));
+    list.add(Expressions.call(BeamEnumerableConverter.class, "toEnumerable", node));
     return implementor.result(physType, list.toBlock());
   }
 
+  public static Enumerable<Object> toEnumerable(BeamRelNode node) {
+    final PipelineOptions options = createPipelineOptions(node.getPipelineOptions());
+    return toEnumerable(options, node);
+  }
+
+  private static PipelineOptions createPipelineOptions(Map<String, String> map) {
+    final String[] args = new String[map.size()];
+    int i = 0;
+    for (Map.Entry<String, String> entry : map.entrySet()) {
```

Comment:
It really seems like `PipelineOptionsFactory` should own this... in fact this makes sense as the primary entry point there, with `--` stripping being just some convenience method.

Reply:
I looked into this and spent a few hours prototyping, but gave up after finding the scope of work involved is non-trivial. `PipelineOptionsFactory` will require a significant refactor to get this done right.